sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -15,74 +18,121 @@ internal class ManagedIdentityClient
     {
         private const string AuthenticationResponseInvalidFormatError = "Invalid response, the authentication response was not in the expected format.";
         private const string MsiEndpointInvalidUriError = "The environment variable MSI_ENDPOINT contains an invalid Uri.";
+        internal const string MsiUnavailableError = "No managed identity endpoint found.";
+        internal const string IdentityUnavailableError = "The requested identity has not been assigned to this resource.";
 
         // IMDS constants. Docs for IMDS are available here https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http
         private static readonly Uri s_imdsEndpoint = new Uri("http://169.254.169.254/metadata/identity/oauth2/token");
+        private static readonly IPAddress s_imdsHostIp = IPAddress.Parse("169.254.169.254");
+        private const int s_imdsPort = 80;
         private const string ImdsApiVersion = "2018-02-01";
-        private const int ImdsAvailableTimeoutMs = 500;
+        private const int ImdsAvailableTimeoutMs = 1000;
 
         // MSI Constants. Docs for MSI are available here https://docs.microsoft.com/en-us/azure/app-service/overview-managed-identity
         private const string AppServiceMsiApiVersion = "2017-09-01";
 
-        private static readonly SemaphoreSlim s_initLock = new SemaphoreSlim(1, 1);
-        private static MsiType s_msiType;
-        private static Uri s_endpoint;
+        private static readonly SemaphoreSlim _initLock = new SemaphoreSlim(1, 1);
+        private MsiType _msiType;
+        private Uri _endpoint;
 
         private readonly CredentialPipeline _pipeline;
 
         protected ManagedIdentityClient()
         {
         }
 
-        public ManagedIdentityClient(CredentialPipeline pipeline)
+        public ManagedIdentityClient(CredentialPipeline pipeline, string clientId = null)
         {
             _pipeline = pipeline;
+
+            ClientId = clientId;
         }
 
-        public virtual AccessToken Authenticate(MsiType msiType, string[] scopes, string clientId, CancellationToken cancellationToken)
+        protected string ClientId { get; }
+
+        public virtual ExtendedAccessToken Authenticate(string[] scopes, CancellationToken cancellationToken)
         {
-            using Request request = CreateAuthRequest(msiType, scopes, clientId);
+            MsiType msiType = GetMsiType(cancellationToken);
+
+            // if msi is unavailable or we were unable to determine the type return CredentialUnavailable exception that no endpoint was found
+            if (msiType == MsiType.Unavailable || msiType == MsiType.Unknown)
```

Comment:
Why would we return Unknown from GetMsiType?

Reply:
Unknown should not be returned. However, since `GetMsiType` returns the member variable `_msiType` I have the check for `MsiType.Unknown` I have this check to guard against any possible races with other setters of this field.